packages/blocks/shuffle/src/components/add-entities-dialog.tsx

```diff
@@ -0,0 +1,205 @@
+import { Entity, EntityType, GraphBlockHandler } from "@blockprotocol/graph/.";
+import {
+  Dialog,
+  DialogTitle,
+  DialogActions,
+  FormControl,
+  FormGroup,
+  FormControlLabel,
+  Checkbox,
+  formControlLabelClasses,
+  experimental_sx as sx,
+  styled,
+} from "@mui/material";
+import { Button } from "@hashintel/hash-design-system";
+import {
+  useImperativeHandle,
+  forwardRef,
+  ForwardRefRenderFunction,
+  useState,
+} from "react";
+import { v4 as uuid } from "uuid";
+import { Items } from "../shuffle";
+import { getEntityLabel } from "../utils";
+
+const SFormControlLabel = styled(FormControlLabel)(
+  sx({
+    width: "100%",
+
+    [`.${formControlLabelClasses.label}`]: {
+      overflow: "hidden",
+      whiteSpace: "nowrap",
+      textOverflow: "ellipsis",
+    },
+  }),
+);
+
+export interface AddEntitiesDialogRef {
+  show: () => void;
+}
+
+interface AddEntitiesDialogProps {
+  entityTypes: EntityType[];
+  blockEntityId: string;
+  onAddEntityItems: (items: Items) => void;
+  graphService?: GraphBlockHandler | null;
+}
+
+const _AddEntitiesDialog: ForwardRefRenderFunction<
+  AddEntitiesDialogRef,
+  AddEntitiesDialogProps
+> = ({ entityTypes, blockEntityId, onAddEntityItems, graphService }, ref) => {
+  const [open, setOpen] = useState(false);
+  const [selectedEntityType, setSelectedEntityType] = useState<EntityType>();
+  const [entityList, setEntityList] = useState<Entity[]>([]);
+  // `selections` is a object that stores checkbox `checked` state for each entity.
+  const [selections, setSelections] = useState<boolean[]>([]);
+
+  const handleClose = () => setOpen(false);
+
+  /**
+   * Defined a custom ref, so the parent can simply call `dialogRef.show()`,
+   * which also resets the state to it's initial value when modal is opened,
+   * instead of using an `useEffect` that listens modal's state
+   */
+  useImperativeHandle(
+    ref,
+    () => ({
+      show: () => {
+        setOpen(true);
+        setSelectedEntityType(undefined);
+        setEntityList([]);
+      },
+    }),
+    [],
+  );
```

Comment:
I personally think is is more logical & close to the developer's intention, because from the parent component we just want to `open the modal`, and pass the modal the dependencies it needs from the parent. I know  `useImperativeHandle` is not that common, but I don't think it's hard to understand either, we just define a open modal function inside the modal, so parent can easily say `open this modal` instead of storing a separate `state` and `handleClose` function. 
_(This does not have to be the case for every modal, some modals maybe need to get closed from outside, but even on that case I still like this approach, since can just add a `close` function to the ref. Or even more the modal state to sth like. I remember using a hook for handling this logic as well. It was sth like `const [modal, showModal, hideModal] = useModal(<ModalContent>)`)_

Reply:
React is a declarative UI framework. The whole point of it is to avoid imperative code. Refs are an imperative escape hatch. It is preferable to avoid using them where possible, and there's no need for them here.  The parent component decides whether the modal is open or not, so when you model the state that way it becomes much easier to understand.

Reply:
I still feel like on **this modal** it's better DX and better readability to trigger the modal opening from a function instead of managing it's state from it's parent. Since this is more like a `dialog` that requires an action, defined by the dialog itself. Like how [confirmation modals](https://ant.design/components/modal/#header) work on Ant Design, or browsers input prompt. But it could achieved in a better way like `const [modal, showModal, hideModal] = useModal(<ModalContent/>)` instead of `useImperativeHandle` with `useRef`. I think we can discuss this again in future if we have more confirmation type of modals in future. I used an api like `useModal` on previous projects, and it was really easy to work with, and easy to read.

Updated this modal to declarative way here: https://github.com/hashintel/hash/pull/1033/commits/750ca5ad37c81e81bbf6a87c2b178d6113e2c4f3